core/src/main/scala/org/apache/spark/deploy/LocalSparkCluster.scala

```diff
@@ -74,6 +74,10 @@ class LocalSparkCluster(
 
   def stop(): Unit = {
     logInfo("Shutting down local Spark cluster.")
+    // SPARK-31922: wait one more second before shutting down rpcEnvs of master and worker,
+    // in order to let the cluster have time to handle the `UnregisterApplication` message.
+    // Otherwise, we could hit "RpcEnv already stopped" error.
+    Thread.sleep(1000)
     // Stop the workers before the master so they don't get upset that it disconnected
     workerRpcEnvs.foreach(_.shutdown())
```

Comment:
additionally, there might be a problem with the comment and the implementation. the code does not really wait for workers to stop before the master. we could rewrite it like:
```
    Seq(workerRpcEnvs, masterRpcEnvs).foreach { rpcEnvArr =>
      rpcEnvArr.foreach(rpcEnv => Utils.tryLog {
        rpcEnv.shutdown()
        rpcEnv.awaitTermination()
      })
      rpcEnvArr.clear()
    }
```

Reply:
It's not really necessary since shutdown is performed in a synchronization way. Therefore, the worker will close the connection to the master firstly. And `awaitTermination()` doesn't make sure everything stops but only the `Dispather`.

Reply:
you may be right about this but this contradicts to the [scaladoc](https://github.com/apache/spark/blob/264b0f36cedacd9a22b45a3e14b2186230432be6/core/src/main/scala/org/apache/spark/rpc/RpcEnv.scala#L119)

Reply:
ok. let's follow the doc.